portal/src/util/stringTemplate.ts

```diff
@@ -0,0 +1,75 @@
+interface StringTemplateParamData {
+  paramName: string;
+  startIndex: number;
+  endIndex: number;
+}
+
+function applyArgumentOnTemplateString(
+  argument: string,
+  template: string,
+  value: string
+): string {
+  return template.replace(new RegExp(`{{[ ]*${argument}[ ]*}}`, "g"), value);
+}
+
+export function renderTemplateString(
+  values: Record<string, string>,
+  template: string
+): string {
+  return Object.entries(values).reduce((result, [key, value]) => {
+    return applyArgumentOnTemplateString(key, result, value);
+  }, template);
+}
+
+export function parseTemplateString(
```

Comment:
This function seems rather complicated for what we need.
I suggest try to hook into template string mechanism for this. Something like:
```ts
function template<ArgKey extends string>(parts: readonly string[], ...args: readonly ArgKey[]) {
    // should receive { parts: ["templates/", "/messages/test.html"], args: ["locale"] }
}
const localeTemplate = template`templates/${'locale'}/messages/test.html`;
```

Reply:
@kiootic I understand this can be used for rendering the string from template, but not sure how to use this for parsing string with template

Reply:
IIRC `parts.length` is always `args.length + 1`, so:
- for each argument `arg` with index `i` in args:
    - check if the input starts with `parts[i]`: if so remove it; otherwise reject the input.
    - find next occurrence of `parts[i+1]` in input: if exists, assign the substring between the two parts to `arg`; otherwise reject the input.

Reply:
I see, using this to replace first part of `parseTemplateString`. So we are using this object
```
{ parts: ["templates/", "/messages/test.html"], args: ["locale"] }
```
as template? @kiootic

Reply:
I'm thinking of returning the parser and renderer as part of tagged template function, something like:
```ts
interface ResourcePath<Arg> {
    parse(path: string): Record<Arg, string> | null;
    render(args: Record<Arg, string>): string;
}
function resourcePath<Arg extends string>(parts: readonly string[], ...args: readonly Arg[]): ResourcePath<Arg> {
    // ...
}

const localizedEmailMessageHTML = resourcePath`templates/${"locale"}/messages/${"type"}_email.html`;
localizedMessageHTML.render({ locale: "en", type: "forgot_password" });
```

Reply:
I see, will implement this